mlflow/store/artifact/optimized_s3_artifact_repo.py

```diff
@@ -0,0 +1,279 @@
+        else:
+            return None
+
+    def _upload_file(self, s3_client, local_file, bucket, key):
+        extra_args = {}
+        guessed_type, guessed_encoding = guess_type(local_file)
+        if guessed_type is not None:
+            extra_args["ContentType"] = guessed_type
+        if guessed_encoding is not None:
+            extra_args["ContentEncoding"] = guessed_encoding
+        environ_extra_args = self.get_s3_file_upload_extra_args()
+        if environ_extra_args is not None:
+            extra_args.update(environ_extra_args)
+        s3_client.upload_file(Filename=local_file, Bucket=bucket, Key=key, ExtraArgs=extra_args)
+
+    def log_artifact(self, local_file, artifact_path=None):
+        artifact_file_path = os.path.basename(local_file)
+        if artifact_path:
+            artifact_file_path = posixpath.join(artifact_path, artifact_file_path)
+        self._upload_to_cloud(
+            cloud_credential_info=self._get_s3_client(),
+            src_file_path=local_file,
+            artifact_file_path=artifact_file_path,
+        )
+
+    def _get_write_credential_infos(self, remote_file_paths):
+        """
+        Instead of returning ArtifactCredentialInfo objects, we instead return a list of initialized
+        S3 client. We do so because S3 clients cannot be instantiated within each thread.
+        """
+        return [self._get_s3_client() for _ in remote_file_paths]
+
+    def _upload_to_cloud(self, cloud_credential_info, src_file_path, artifact_file_path):
+        dest_path = posixpath.join(self.bucket_path, artifact_file_path)
+        key = posixpath.normpath(dest_path)
+        if (
+            MLFLOW_ENABLE_MULTIPART_UPLOAD.get()
+            and os.path.getsize(src_file_path) > _MULTIPART_UPLOAD_CHUNK_SIZE
+        ):
+            self._multipart_upload(cloud_credential_info, src_file_path, self.bucket, key)
+        else:
+            self._upload_file(cloud_credential_info, src_file_path, self.bucket, key)
+
+    def _multipart_upload(self, cloud_credential_info, local_file, bucket, key):
+        # Create multipart upload
+        s3_client = cloud_credential_info
+        response = s3_client.create_multipart_upload(Bucket=bucket, Key=key)
+        upload_id = response["UploadId"]
+
+        # Create presigned URL for each part
+        num_parts = math.ceil(os.path.getsize(local_file) / _MULTIPART_UPLOAD_CHUNK_SIZE)
+
+        def _get_presigned_upload_part_url(part_number):
+            return s3_client.generate_presigned_url(
+                "upload_part",
+                Params={
+                    "Bucket": bucket,
+                    "Key": key,
+                    "UploadId": upload_id,
+                    "PartNumber": part_number,
+                },
+            )
+
+        # define helper functions for uploading data
+        def _upload_part(part_number, local_file, start_byte, size):
+            presigned_url = _get_presigned_upload_part_url(part_number)
```

Comment:
```suggestion
            presigned_url = s3_client.generate_presigned_url(...)
```

can we run `generate_presigned_url` here instead of creating another function?

Reply:
Done